File: leetcode/c/1318-minimum-flips-to-make-a-or-b-equal-to-c.c
// Author: RT
// Date: 2023-06-07T04:02:13.047Z
// URL: https://leetcode.com/problems/minimum-flips-to-make-a-or-b-equal-to-c/description/

int minFlips(int a, int b, int c)
{
    // if bit at c is unset, both set bit(s) in a and b needs to be unset
    // i.e., if bit is set on (a & b) but not c, we need to double count.
    int diff_mask = (a | b) ^ c;
    return __builtin_popcount(diff_mask) + __builtin_popcount(a & b & diff_mask);
}

File: leetcode/python/1318-minimum-flips-to-make-a-or-b-equal-to-c.py
# Author: RT
# Date: 2023-06-07T03:42:13.799Z
# URL: https://leetcode.com/problems/minimum-flips-to-make-a-or-b-equal-to-c/


class Solution:
    def minFlips(self, a: int, b: int, c: int) -> int:
        ans = 0
        while a or b or c:
            ba = a & 1
            bb = b & 1
            bc = c & 1
            if ba | bb != bc:
                if bc:
                    ans += 1
                else:
                    ans += bool(ba) + bool(bb)

            a >>= 1
            b >>= 1
            c >>= 1

        return ans
